coding_guidelines.md

```diff
@@ -0,0 +1,500 @@
+
+### [Don't use Arguments](#dont-use-arguments)
+To clarify, your functions can have arguments, but you should never name a parameter `arguments` as this will take precedence over the similarly named object on the function scope, and you don't want to do that.
+
+### [Default Parameters](#default-parameters)
+Instead of having mutating function arguments, consider using default parameters.
+```typescript
+// Worst implementation
+function handleThings(opts): void {
+  // No! We shouldn't mutate function arguments.
+  // Double bad: if opts is falsy it'll be set to an object which may
+  // be what you want but it can introduce subtle bugs.
+  opts = opts || {};
+}
+
+// Not much better
+function handleThings(opts): void {
+  if(opts === void 0) {
+    opts = {};
+  }
+}
+
+// Correct way
+
+function handleThings(opts = {}): void {
+
+}
+```
+
+### [Function Signature](#function-signature)
+When creating a stand-alone function for usage across different files there are some rules that must be followed.
+- Function name
+  - descriptive & precise
+  - camelCased
+  - non-abbreviated
+- Function variables
+  - descriptive & precise
+  - camelCased
+  - [destructured](#when-should-you-use-destructuring)
+  - typed
+- Function returntype
+  - always defined
+
+```typescript
+
+// Silly example, obviously don't do this.
+function hepp(a: number, b: number) {
+  return Math.floor(Math.random() * (max - min) + min);
+}
+
+// Small improvement
+function randomValue(min: number, max: number) {
+  return Math.round(Math.random() * (max - min) + min);
+}
+
+// Eureka
+function randomNumberBetween(min: number, max: number): number {
+  return Math.round(Math.random() * (max - min) + min);
+}
+```
+
+### [Class Functions](#class-functions)
+Functions defined in class should follow the strict rules of function signatures. In addition to the general rules, class functions require defining the access level of the function.
+```typescript
+class Thermostat {
+  
+  private _targetTemperature: number;
+  private _maximumTemperatureLimit: number;
+  private _minimumTemperatureLimit: number;
+
+  constructor(
+    targetTemperature: number = 22,
+    minimumTemperatureLimit: number = 5,
+    maximumTemperatureLimit: number = 35
+  ) {
+    this._targetTemperature = targetTemperature;
+    this.minimumTemperatureLimit = minimumTemperatureLimit;
+    this.maximumTemperatureLimit = maximumTemperatureLimit;
+  }
+
+  // Bad function
+  change(num: number) {
+    this._targetTemperature += num;
+    if(this._targetTemperature > this._maximumTemperaturLimit) {
+      this._targetTemperature = this._maximumTemperatureLimit;
+    }
+    else if(this._targetTemperature < this._minimumTemperaturLimit) {
+      this._targetTemperature = this._minimumTemperatureLimit;
+    }
+    return this._targetTemperature;
+  }
+
+  // Good function
+  public getTargetTemperatureAfterUpdatingTemperatureBy(degrees: number): number {
```

Comment:
I disagree. What does 'update temperature by' mean? Increase? I think `increaseTemperatur(deltaDegrees: number): number` would be good...

However, I'm not sure if I agree that we should be this granular in our code style definitions.

Reply:
I am saying, either a function is:
A) Super Granular
B) Documented with /** **/